src/Voron/Impl/FreeSpace/FreeSpaceHandling.cs

```diff
@@ -185,35 +198,57 @@ public void Clear()
             }
         }
 
-
         private long? TryFindSmallValue(LowLevelTransaction tx, FixedSizeTree freeSpaceTree, FixedSizeTree.IFixedSizeIterator it, int num)
         {
             do
             {
-                var current = new StreamBitArray(it.CreateReaderForCurrent());
-
+                long currentSectionId = it.CurrentKey;
                 long? page;
-                if (current.SetCount < num)
+
+                if (_maxConsecutiveRangePerSection.TryGetValue(currentSectionId, out var known) && known.Max <= num)
                 {
-                    if (TryFindSmallValueMergingTwoSections(tx, freeSpaceTree, it.CurrentKey, num, current, out page))
+                    if (_maxConsecutiveRangePerSection.TryGetValue(currentSectionId + 1, out var knownNext))
+                    {
+                        if (known.EndBits + knownNext.StartBits < num)
```

Comment:
With the default settings `EndBits` (2048) + `StartBits` (2048) is `4096`. Is this check valid then?

Reply:
Yes. `EndBits` = 2048 and `StartBits` = 2048 is when we don't know the actual value. That means that we need to check both sections in `TryFindSmallValueMergingTwoSections` to find out if we have enough free space.